### PR TITLE
Cors middleware extension

### PR DESCRIPTION
**Pull request by k-nasa**
## Description

Cors middleware extension. Added validation of origin header.

## Motivation and Context

https://github.com/rustasync/tide/pull/262#pullrequestreview-242949154

## How Has This Been Tested?

By test code, also by example.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/rustasync/tide/blob/master/.github/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by prasannavl** [comment]
@k-nasa - Just glanced through - so neatly done! Thanks a ton :smile. Will go through in detail shortly.

**Comment by k-nasa** [comment]
Surely!

**Comment by k-nasa** [comment]
@prasannavl I fixed it

File: tide-cors/src/middleware.rs [hunk 1]
```diff
@@ -28,11 +28,53 @@ pub struct CorsMiddleware {
...
 }
 
+/// allow_origin enum
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub enum CorsOrigin {
+    /// Wildcard. Accept all origin requests
+    Any,
+    /// Set one allow_origin
+    Exact(String),
+    /// Set some allow_origin
```
Comment (yoshuawuyts) [inline tide-cors/src/middleware.rs]:
```suggestion
    /// Set multiple allow_origin targets
```

File: tide-cors/src/middleware.rs [hunk 2]
```diff
@@ -28,11 +28,53 @@ pub struct CorsMiddleware {
...
     expose_headers: Option<HeaderValue>,
     max_age: HeaderValue,
 }
 
+/// allow_origin enum
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub enum CorsOrigin {
+    /// Wildcard. Accept all origin requests
+    Any,
+    /// Set one allow_origin
```
Comment (yoshuawuyts) [inline tide-cors/src/middleware.rs]:
```suggestion
    /// Set a single allow_origin target
```

**Comment by yoshuawuyts** [comment]
@k-nasa I think all that's missing for this PR now is a rebase, and then we should be good to merge! (:

**Comment by k-nasa** [comment]
@yoshuawuyts I merged master!

**Pull request merged**